File: 08. Trees/01. Binary Tree/03. Postorder Traversal/main.py
class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

# Postorder Traversal (Recursive)
def postorder_traversal_recursive(root, result):
    if not root:
        return
    postorder_traversal_recursive(root.left, result)
    postorder_traversal_recursive(root.right, result)
    result.append(root.val)

# Postorder Traversal (Iterative)
def postorder_traversal_iterative(root):
    if not root:
        return []
    
    result = []
    stack1, stack2 = [root], []

    while stack1:
        node = stack1.pop()
        stack2.append(node)
        if node.left:
            stack1.append(node.left)
        if node.right:
            stack1.append(node.right)

    while stack2:
        result.append(stack2.pop().val)

    return result

# Main function
def main():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)

    recursive_result = []
    postorder_traversal_recursive(root, recursive_result)
    print("Recursive Postorder Traversal:", recursive_result)

    iterative_result = postorder_traversal_iterative(root)
    print("Iterative Postorder Traversal:", iterative_result)

# Run the main function
main()

File: 08. Trees/01. Binary Tree/03. Postorder Traversal/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// Postorder Traversal (Recursive)
function postorderTraversalRecursive(root, result) {
    if (!root) return;
    postorderTraversalRecursive(root.left, result);
    postorderTraversalRecursive(root.right, result);
    result.push(root.val);
}

// Postorder Traversal (Iterative)
function postorderTraversalIterative(root) {
    let result = [];
    if (!root) return result;

    let stack1 = [], stack2 = [];
    stack1.push(root);

    while (stack1.length > 0) {
        let node = stack1.pop();
        stack2.push(node);

        if (node.left) stack1.push(node.left);
        if (node.right) stack1.push(node.right);
    }

    while (stack2.length > 0) {
        result.push(stack2.pop().val);
    }

    return result;
}

// Main function
function main() {
    let root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.left = new TreeNode(4);
    root.left.right = new TreeNode(5);
    root.right.left = new TreeNode(6);
    root.right.right = new TreeNode(7);

    let recursiveResult = [];
    postorderTraversalRecursive(root, recursiveResult);
    console.log("Recursive Postorder Traversal:", recursiveResult);

    let iterativeResult = postorderTraversalIterative(root);
    console.log("Iterative Postorder Traversal:", iterativeResult);
}

// Run the main function
main();
